embulk-core/src/main/java/org/embulk/plugin/maven/MavenPluginSource.java

```diff
@@ -0,0 +1,168 @@
+    {
+        final String category;
+        if (InputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "input";
+        } else if (OutputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "output";
+        } else if (ParserPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "parser";
+        } else if (FormatterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "formatter";
+        } else if (DecoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "decoder";
+        } else if (EncoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "encoder";
+        } else if (FilterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "filter";
+        } else if (GuessPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "guess";
+        } else if (ExecutorPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "executor";
+        } else {
+            // unsupported plugin category
+            throw new PluginSourceNotMatchException("Plugin interface " + pluginInterface + " is not supported.");
+        }
+
+        if (pluginType.getSourceType() != PluginSource.Type.MAVEN) {
+            throw new PluginSourceNotMatchException();
+        }
+        final MavenPluginType mavenPluginType = (MavenPluginType) pluginType;
+
+        final PluginClassLoaderFactory pluginClassLoaderFactory =
+            this.injector.getInstance(PluginClassLoaderFactory.class);
+
+        final MavenArtifactFinder mavenArtifactFinder;
+        try {
+            mavenArtifactFinder = MavenArtifactFinder.create(getLocalMavenRepository());
+        }
+        catch (MavenRepositoryNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Path jarPath;
+        try {
+            jarPath = mavenArtifactFinder.findMavenArtifactJar(
+                mavenPluginType.getGroup(),
+                "embulk-" + category + "-" + mavenPluginType.getName(),
+                mavenPluginType.getVersion());
+        }
+        catch (MavenArtifactNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Class pluginMainClass;
+        try (JarPluginLoader loader = JarPluginLoader.load(jarPath, pluginClassLoaderFactory)) {
+            pluginMainClass = loader.getPluginMainClass();
+        }
+        catch (InvalidJarPluginException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        if (!pluginInterface.isAssignableFrom(pluginMainClass)) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() + "\" is not a valid " + category + " plugin.");
+        }
+
+        final Object pluginMainObject;
+        try {
+            pluginMainObject = pluginMainClass.newInstance();
```

Comment:
I was confusing.. We should've checked `JRubyPluginSource` instead of `InjectedPluginSource`. Because `MavenPluginSource` is similar to `JRubyPluginSource`. I confirmed how Embulk loads S3InputPlugin in Embulk. 

1. lib/embulk/runner.rb calls `EmbulkEmbed#run`.
2. calls `BulkLoader#run`, and `doRun`.
3. calls `Exec#newPlugin`, and `PluginManager#newPlugin`.
4. calls `JRubyPluginSource#newPlugin`
5. calls `lib/embulk/plugin.rb`, Ruby's `PluginManager#new_java_input`.
6. calls `lib/embulk/plugin_registry.rb`'s `lookup`, `search`, and `require_and_show`.
7. calls `lib/embulk/input/s3.rb`'s `JavaPlugin#register_input`. 
8. calls `lib/embulk/plugin.rb` `register_java_input`, `register_java_plugin`.
9. calls `found.last.from_java(klass)`, file_input_plugin.rb's `from_java`.
10. calls `@registries[category].register(type, adapted)` to register it.
11. After `lookup` method call, plugin.rb's `new_java_input` calls `new_java` to create `FileInputRunner`.